net/libfab/libfab.c

```diff
@@ -0,0 +1,2963 @@
+	if (rc != FI_SUCCESS)
+		return M0_ERR(rc);
+
+	ev.events = EPOLLIN;
+	ev.data.ptr = ctx;
+	rc = epoll_ctl(tm->ftm_epfd, EPOLL_CTL_ADD, fd, &ev);
+
+	return M0_RC(rc);
+}
+
+/**
+ * Return the pointer to the active endpoint from the m0_fab__ep struct
+ */
+static inline struct m0_fab__active_ep *libfab_aep_get(struct m0_fab__ep *ep)
+{
+	return (ep->fep_listen == NULL) ? ep->fep_aep : ep->fep_listen->pep_aep;
+}
+
+/**
+ * Returns true if verbs provider is selected else false.
+ */
+static inline bool libfab_is_verbs(struct m0_fab__tm *tm)
+{
+	return (!strcmp(tm->ftm_fab->fab_fi->fabric_attr->prov_name, "verbs"));
+}
+
+/**
+ * Add bulk operation into transfer machine bulk operation list.
+ */
+static int libfab_bulklist_add(struct m0_fab__tm *tm, struct m0_fab__buf *fb,
+				struct m0_fab__active_ep *aep)
+{
+	struct m0_fab__bulk_op *op;
+
+	M0_ALLOC_PTR(op);
+	if (op == NULL)
+		return -ENOMEM;
+	op->fbl_aep = aep;
+	op->fbl_buf = fb;
+	fb->fb_bulk_op = op;
+	fab_bulk_tlink_init_at_tail(op, &tm->ftm_bulk);
+
+	return M0_RC(0);
+}
+
+/** 
+ * This function will calculate the number of work requests that will be 
+ * generated the for the bulk transfer operation (read/write) on the
+ * net-buffer.
+ * The wr_cnt is returned
+ */
+static uint32_t libfab_wr_cnt_get(struct m0_fab__buf *fb)
+{
+	struct fi_rma_iov *r_iov;
+	m0_bcount_t       *v_cnt = fb->fb_nb->nb_buffer.ov_vec.v_count;
+	m0_bcount_t        xfer_len = 0;
+	m0_bcount_t        chunk;
+	uint32_t           loc_sidx = 0;
+	uint32_t           rem_sidx = 0;
+	uint32_t           loc_soff = 0;
+	uint32_t           rem_soff = 0;
+	uint32_t           loc_slen;
+	uint32_t           rem_slen;
+	uint32_t           wr_cnt = 0;
+
+	M0_ENTRY("loc_buf=%p q=%d loc_seg=%d rem_buf=0x%"PRIx64" rem_seg=%d",
+		 fb, fb->fb_nb->nb_qtype, fb->fb_nb->nb_buffer.ov_vec.v_nr, 
+		 fb->fb_rbd->fbd_bufptr, (int)fb->fb_rbd->fbd_iov_cnt);
+	M0_PRE(fb->fb_rbd != NULL);
+
+	r_iov = fb->fb_riov;
+
+	while(xfer_len < fb->fb_nb->nb_length) {
+		M0_ASSERT(rem_sidx <= fb->fb_rbd->fbd_iov_cnt);
+		loc_slen = v_cnt[loc_sidx] - loc_soff;
+		rem_slen = r_iov[rem_sidx].len - rem_soff;
+		
+		chunk = min64u(loc_slen, rem_slen);
+
+		if (loc_slen > rem_slen) {
+			rem_sidx++;
+			rem_soff = 0;
+			loc_soff += chunk;
+		} else {
+			loc_sidx++;
+			loc_soff = 0;
+			rem_soff += chunk;
+			if(rem_soff >= r_iov[rem_sidx].len) {
+				rem_sidx++;
+				rem_soff = 0;
+			}
+		}
+		wr_cnt++;
+		xfer_len += chunk;
+	}
+	return wr_cnt;
+}
+
+/**
+ * Process bulk operation from transfer machine bulk operation list.
+ */
+static void libfab_bulk_buf_process(struct m0_fab__tm *tm)
+{
+	struct m0_fab__bulk_op *op;
+	uint32_t                available;
+
+	m0_tl_for(fab_bulk, &tm->ftm_bulk, op) {
+		/* Keeping 64 entries on CQ for rpc messages */
+		available = (FAB_MAX_TX_CQ_EV - 64) - op->fbl_aep->aep_bulk_cnt;
+
+		if (op->fbl_buf->fb_wr_cnt <= available &&
+		    op->fbl_aep->aep_tx_state == FAB_CONNECTED) {
```

Comment:
Please add comment (with required modification) as we discussed! 
_In one of the ST, the IOservice process is killed midway during the test (to simulate a crash), in such cases we are holding the bulk transfers till the connection is reestablished. Thats why the check for aep_tx_state == FAB_CONNECTED_

Reply:
Comment added